pkg/preflight/aws.go

```diff
@@ -0,0 +1,133 @@
+		return nil
+	}
+
+	// Initialize aws client for cloud drive permission checks
+	// TODO: use provided credentials to create a client
+	client, err := awsops.NewClient()
+	if err != nil {
+		logrus.Infof("JLIAO: err: %v", err)
+		return err
+	}
+	a.ops = client
+
+	instance, err := client.InspectInstance(client.InstanceID())
+	if err != nil {
+		return err
+	}
+	a.zone = instance.Zone
+	logrus.WithFields(logrus.Fields{
+		"instance": instance,
+	}).Infof("JLIAO: get instance")
+	return nil
+}
+
+func (a *aws) getEC2VolumeTemplate() *ec2.Volume {
+	volTypeGp2 := opsworks.VolumeTypeGp2
+	volSize := int64(1)
+	return &ec2.Volume{
+		Size:             &volSize,
+		VolumeType:       &volTypeGp2,
+		AvailabilityZone: &a.zone,
+	}
+}
+
+func (a *aws) CheckCloudDrivePermission() error {
+	// Only init the aws client when needed
+	if err := a.initCloudOps(); err != nil {
+		return err
+	}
+	// check the permission here by doing dummy drive operations
+	logrus.Infof("JLIAO: doing aws permisison check")
+
+	// Create volume
+	// TODO: add more identifier to labels like cluster name
+	v, err := a.ops.Create(a.getEC2VolumeTemplate(), map[string]string{"name": "px-preflight-check"})
+	if err != nil {
+		// Volume creation doesn't take volume name, so it's guaranteed there's no conflict
+		logrus.Errorf("JLIAO: failed to create aws volume: %v", err)
+		return err
+	}
+
+	logrus.WithFields(logrus.Fields{
+		"volume": v,
+	}).Infof("JLIAO: volume created")
+
+	// Attach volume
+	vol := v.(*ec2.Volume)
+	_, err = a.ops.Attach(*vol.VolumeId, nil)
+	if err != nil {
+		// Operator container doesn't have access to /dev/ directory on the host
+		// unable to map volume vol-03227127b9b7b442e with block device mapping /dev/xvdp to an actual device path on the host
+		errGetPath := fmt.Sprintf("unable to map volume %s with block device mapping", *vol.VolumeId)
+		errAlreadyAttached := fmt.Sprintf("%s is already attached to an instance", *vol.VolumeId)
+		if !strings.Contains(err.Error(), errGetPath) && !strings.Contains(err.Error(), errAlreadyAttached) {
+			logrus.WithError(err).Errorf("JLIAO: error attach volume")
+			return err
+		}
+
+		// Try to get the volume attachment info
+		volumes, err := a.ops.Inspect([]*string{vol.VolumeId})
+		if err != nil || len(volumes) != 1 {
+			logrus.WithError(err).Errorf("JLIAO: error inspect volume")
+			return fmt.Errorf("failed to inspect volume %s: %v", *vol.VolumeId, err)
+		}
+		vol := volumes[0].(*ec2.Volume)
+		logrus.WithFields(logrus.Fields{
+			"vol": vol,
+		}).Infof("JLIAO: vol inspected")
+		if len(vol.Attachments) != 1 || *vol.Attachments[0].State != ec2.VolumeAttachmentStateAttached {
+			logrus.WithFields(logrus.Fields{
+				"volume": vol,
+			}).Errorf("JLIAO: volume not attached")
+			return fmt.Errorf("JLIAO: volume not attached")
+		}
+		logrus.WithFields(logrus.Fields{
+			"volume": vol,
+		}).Infof("JLIAO: volume attached")
+	}
+
+	// Detach
+	if err := a.ops.Detach(*vol.VolumeId); err != nil {
```

Comment:
This flow makes me nervous: if we fail partway through, we leave an EBS volume still attached to the volume, or still created at the very least, and those costs can add up very quickly (I know from homelab experience 🙃 ).

Can we put the Detach and Delete inside a defer block? That way we'll always try to clean up after ourselves. It may be easier to split it up into multiple methods so you can have an individual defer on each one (have the creation process be one method, which will defer deletion, and then the attach method will be separate and defer detachment).